### PR TITLE
fix(patterns): type guards

### PR DESCRIPTION
**Pull request by erights**
Subset of https://github.com/endojs/endo/pull/1712 that does not break compat.

Typing and validation for `InterfaceGuard`s, `MethodGuard`s, and `AwaitArgGuard`s.

### CODE REVIEW

**Comment by erights** [comment]
@michaelfig , making you a reviewer because I'm staging my interface-opt-out experiment on this. Let me know if you see a problem with that. I so, I can remove the dependency, but I'd rather not.

**Comment by erights** [comment]
Note that I'd prefer to stage on https://github.com/endojs/endo/pull/1712 if that were feasible. But I'm scared of the compat break. Eager for your feedback on that too.

**Comment by erights** [comment]
@turadg Making you a reviewer as I know you've done previous work typing guards. I don't know how that relates to the contents of this PR. What should I look at?

**Comment by michaelfig** [comment]
> Note that I'd prefer to stage on #1712 if that were feasible. But I'm scared of the compat break. Eager for your feedback on that too.

I believe the breaking change in #1712 won't impact any deployed code.  IIUC, that's the change of guards from copyRecords to tagged payloads.  I can't think of any instance where we put guards in object state or Stores in general, or send them over the network.

If that seems correct to you, I'd encourage you to make the change and stage on #1712.  I had just one review comment on #1712, and would be happy to approve after you consider that question.

**Comment by erights** [comment]
Since #1712 is staged on this, even if we decide we can merge #1712, it doesn't hurt to first merge this PR, once it meets your approval.

File: packages/patterns/src/patterns/patternMatchers.js
Comment (michaelfig) [reply packages/patterns/src/patterns/patternMatchers.js:523]: To aid in understanding, can you briefly describe the problem here, or make an issue for it?

File: packages/patterns/src/patterns/patternMatchers.js
Comment (michaelfig) [inline packages/patterns/src/patterns/patternMatchers.js:960]: Does this comment still apply?

File: packages/patterns/src/patterns/patternMatchers.js
Comment (michaelfig) [inline packages/patterns/src/patterns/patternMatchers.js:1731]: Would it be worthwhile to try to push the `argPattern` semantics as far down as possible without breaking compatibility?

```suggestion
 * @param {Pattern} argPattern
 * @returns {AwaitArgGuard}
 */
const makeAwaitArgGuard = argPattern => {
  /** @type {AwaitArgGuard} */
  const result = harden({
    klass: 'awaitArg',
    argGuard: argPattern,
  });
  assertAwaitArgGuard(result);
  return result;
};
```

**Review by michaelfig — APPROVED** [review]
Approval based on your addressing my comments.

**Comment by erights** [comment]
> > Note that I'd prefer to stage on #1712 if that were feasible. But I'm scared of the compat break. Eager for your feedback on that too.
> 
> I believe the breaking change in #1712 won't impact any deployed code. IIUC, that's the change of guards from copyRecords to tagged payloads. I can't think of any instance where we put guards in object state or Stores in general, or send them over the network.
> 
> If that seems correct to you, I'd encourage you to make the change and stage on #1712. I had just one review comment on #1712, and would be happy to approve after you consider that question.

AFAICT, you're correct re durable state and network messages. However, #1712 seems incompat with some of the code in current agoric-sdk. For example:

image

File: packages/patterns/src/patterns/patternMatchers.js
Comment (erights) [reply packages/patterns/src/patterns/patternMatchers.js:523]: The argument of `copyMapKeySet` is supposed to be a CopyMap. At this location, both `patt` and `specimen` are known to be CopyMaps. The payload of a CopyMap is a pair of a keys array and a corresponding values array. Such a pair is not itself a valid CopyMap.

This fix does do redundant work, in that `copyMapKeySet` reverifies that its argument is a CopyMap, and after that only pays attention to the payload anyway. But with the memoes we already have, that should only be a tiny bit of redundant work. Avoid it would make abstraction boundaries a bit messier.

Perhaps one day our static type checking would be precise enough to have flagged this error ;) .

Comment (erights) [reply packages/patterns/src/patterns/patternMatchers.js:523]: Resolving. But let me know if you think it is worth opening a closing an issue for this instead.

Comment (michaelfig) [reply packages/patterns/src/patterns/patternMatchers.js:523]: > opening and closing an issue

No need.  Thanks for the explanation!

File: packages/patterns/src/patterns/patternMatchers.js
Comment (erights) [reply packages/patterns/src/patterns/patternMatchers.js:535]: Filed at https://github.com/endojs/endo/issues/1727 and assigned to me. Resolving here.

File: packages/patterns/src/patterns/patternMatchers.js
Comment (erights) [reply packages/patterns/src/patterns/patternMatchers.js:960]: The comment is no longer informative. Gone.

File: packages/patterns/src/patterns/patternMatchers.js
Comment (erights) [reply packages/patterns/src/patterns/patternMatchers.js:1731]: Done

**Pull request merged**
